### ISSUE TITLE
Async inference behaves differently between migrated older models and newly trained models

### ISSUE THREAD

**Issue by aShunSasaki** [huggingface/lerobot#2980]
### Ticket Type

🐛 Bug Report (Something isn't working)

### Environment & System Info

```Shell
- LeRobot: commit at 14a15f90
- Robot: SO-101
- Policy: SmolVLA
- OS: macOS

- the output of `lerobot-info`

- LeRobot version: 0.4.4
- Platform: macOS-15.7.2-arm64-arm-64bit
- Python version: 3.11.9
- Huggingface Hub version: 0.35.3
- Datasets version: 4.1.1
- Numpy version: 2.2.6
- FFmpeg version: 7.1.1
- PyTorch version: 2.7.1
- Is PyTorch built with CUDA support?: False
- Cuda version: N/A
- GPU model: N/A
- Using GPU in script?: <fill in>
- lerobot scripts: ['lerobot-calibrate', 'lerobot-dataset-viz', 'lerobot-edit-dataset', 'lerobot-eval', 'lerobot-find-cameras', 'lerobot-find-joint-limits', 'lerobot-find-port', 'lerobot-imgtransform-viz', 'lerobot-info', 'lerobot-record', 'lerobot-replay', 'lerobot-setup-can', 'lerobot-setup-motors', 'lerobot-teleoperate', 'lerobot-train', 'lerobot-train-tokenizer']
```

### Description

We observe a noticeable difference in task success between Sync (local) inference and Async inference.

In our setup, policies trained with the current version of LeRobot perform reliably under Sync inference, but often fail under Async inference. In contrast, policies trained with older LeRobot versions and then migrated to the current format tend to show much smaller degradation when run with Async inference.

The issue is reproducible across multiple runs of the same task and does not appear to be limited to a single dataset or model checkpoint.

#### Observed behavior

* Sync inference achieves relatively stable task success.
* Async inference shows significantly lower success rates.
* The robot arm usually moves close to the target, but precise actions (e.g. grasping) often fail under Async inference.

A consistent difference we observed is how the policy was trained:

* Policies trained with older LeRobot versions (around commit 483be9a ) tend to retain reasonable performance under Async inference.
* Policies fine-tuned from the same base model using the current version of LeRobot often show severe degradation under Async inference, even when trained on the same dataset migrated to Dataset v3.

#### Policies and datasets used

| Policy         | Training setup                               | Dataset version   | Link                                                                                                                                                                       |
| -------------- | -------------------------------------------- | ----------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Old (migrated) | Fine-tuned with older LeRobot, then migrated | Dataset v2.1      | [https://huggingface.co/yutaro-kimura-acs/so101_pp_blue_box_policy_01_migrated](https://huggingface.co/yutaro-kimura-acs/so101_pp_blue_box_policy_01_migrated)             |
| New            | Fine-tuned with current LeRobot              | Dataset v2.1 → v3 (Migrated with [conversion script](https://huggingface.co/docs/lerobot/lerobot-dataset-v3#migrate-v21--v30)) | [https://huggingface.co/aShunSasaki/so101_pp_blue_box_yk_dataset_policy_v30_cp_20000](https://huggingface.co/aShunSasaki/so101_pp_blue_box_yk_dataset_policy_v30_cp_20000) |

Both policies were fine-tuned from the same base model (`smolvla_base`) using the same dataset:
[https://huggingface.co/datasets/yutaro-kimura-acs/so101_pp_blue_box](https://huggingface.co/datasets/yutaro-kimura-acs/so101_pp_blue_box)

#### Videos

* Old (migrated):

https://github.com/user-attachments/assets/d362ed09-112c-46ec-adec-7058c4b6ea3f

* New:

https://github.com/user-attachments/assets/c3c6fcd8-f169-4324-8aba-9148af304e41

#### Task success rates

| Policy         | Sync inference | Async inference | Notes                                    |
| -------------- | :------------: | :-------------: | ---------------------------------------- |
| Old (migrated) |     80–90%     |       **~50%**      | `input_features` shape: `[3, 720, 1280]` |
| New (current)  |     80–90%     |        **0%**       | `input_features` shape: `[3, 256, 256]`  |

* Robot: SO-101
* Task: Pick-and-place (blue box)
* Policy architecture: SmolVLA (`smolvla_base` fine-tuned)

#### Possible contributing factor

One possible factor we identified is a difference in image preprocessing between Sync and Async inference paths.

In the async path, the camera image appears to be resized to match `policy.config.input_features` before being passed to the model. For policies where `input_features` reflects the dataset storage resolution (e.g. 256×256), this results in the camera image being resized from its native resolution (1280×720) to a square shape.

In contrast, Sync inference does not resize the image at this stage and relies on the model’s internal `resize_with_pad`, which preserves aspect ratio.

This behavior seems related to the image resizing discrepancy discussed in #2475, and may help explain why older migrated policies behave more robustly under Async inference.

#### Configuration differences observed

While investigating this issue, we also compared the policy configuration files between older migrated policies and newly fine-tuned ones.

One notable difference is the `input_features` image shape:

* Older migrated policies use the camera’s native resolution (e.g. `[3, 720, 1280]`)
* Newly fine-tuned policies use the dataset storage resolution (e.g. `[3, 256, 256]`)

This difference appears to affect how images are handled in the async inference path, where the input image is resized to match `input_features` before being passed to the policy.

We are not fully sure whether this alone explains the observed behavior, but it stood out during the comparison.

### Relevant logs or stack trace

```Shell
No explicit runtime errors or stack traces were observed.
```

### Checklist

- [x] I have searched existing tickets to ensure this isn't a duplicate.
- [x] I am using the latest version of the `main` branch.
- [x] I have verified this is not an environment-specific problem.

### Additional Info / Workarounds

This issue may be related to the image preprocessing difference discussed in:

- https://github.com/huggingface/lerobot/issues/2475

That issue focuses on image resizing behavior specifically, while this ticket aims to describe the broader performance gap observed between Sync and Async inference, as well as the difference between older migrated models and newly trained ones.

At the moment, using a policy trained with an older LeRobot version and migrating it to the current format appears to mitigate the issue.

**Comment by mgm136044** [comment]
Hello, I was wondering how you managed to use dataset v2.1.
My dataset is currently v3.0, and to get async inference working properly, it seems I may need to downgrade to v2.1.
Could you share the steps you took to downgrade (e.g., whether you used a conversion script, edited metadata, or re-uploaded the dataset)? Any guidance would be greatly appreciated.

**Comment by aShunSasaki** [comment]
@mgm136044 

At the moment, there isn’t a supported way to downgrade a dataset from v3.0 back to v2.1.

What I did instead was:
1. Use an older version of LeRobot to collect a v2.1 dataset, and train the policy in the same environment.
2. Then switch to the current version of LeRobot and run the following to migrate the trained policy:

```
python -m lerobot.processor.migrate_policy_normalization \
  --pretrained-path your-username/your-old-policy \
  --push-to-hub
```

So the dataset collection and training are done with the old version, and only the policy migration is done with the current version.

**Comment by rylinjames** [comment]
the sync-vs-async gap on smolvla + so-101 is almost always one of two things: (1) action-chunk horizon mismatch (sync replays the full chunk, async often only uses the first k actions then requests a new one, so drift accumulates on precise grasps), or (2) observation timestamp alignment (async server processes a slightly stale obs, old-version checkpoints were trained with more temporal smoothing so they tolerate it). can you check the `n_action_steps` / chunk-consumption setting on the async client vs the sync path? that's the first thing to pin. reflex won't fix the training-time root cause here, but we do provide a deterministic onnx+fp16-trt deploy for smolvla with cos=+1.0 parity, useful for isolating whether async drift is in the server/scheduler or the policy itself. github.com/rylinjames/reflex-vla has good-first-issues tagged, or dm romirj on discord / playindus@gmail.com for collaborator access.

**Comment by helper2424** [comment]
I would rebuild the the async fully. I have started doing it, but didn't finish - check my PRs. On the current project we rebuilt it fully with sync remote commucation and one thread that send peridoically requests for inference.
